### PR TITLE
fix(exec source): emit discarded error when channel is closed

### PR DESCRIPTION
**Pull request by StephenWakely**
Ref #14451 

If the `Sender` is closed when the command thread tries to send it, that event is dropped. Before we were emitting a `trace`. I've updated this to emit an `Error` and update `component_discarded_events_total`.

I can't really think of a realistic situation where this error would occur, but it is a potential so we should emit.

### CODE REVIEW

File: src/internal_events/exec.rs [hunk 1]
```diff
@@ -219,3 +219,25 @@ impl InternalEvent for ExecFailedToSignalChildError<'_> {
...
+        counter!(
+            "component_errors_total", 1,
+            "error_type" => error_type::COMMAND_FAILED,
+            "stage" => error_stage::RECEIVING,
+        );
+        counter!(
+            "component_discarded_events_total", 1,
+            "error_type" => error_type::COMMAND_FAILED,
+            "stage" => error_stage::RECEIVING,
+        );
```
Comment (neuronull) [inline src/internal_events/exec.rs:241]:
```suggestion
          emit!(ComponentEventsDropped::<UNINTENTIONAL> {                                                                                                                                                                                                                                                                                               
              count: 1,                                                                                                                                                                                                                                                                                                                      
              reason,                                                                                                                                                                                                                                                                                             
          });     
```

File: src/internal_events/exec.rs [hunk 2]
```diff
@@ -219,3 +219,25 @@ impl InternalEvent for ExecFailedToSignalChildError<'_> {
...
+
+pub struct ExecChannelClosedError;
+
+impl InternalEvent for ExecChannelClosedError {
+    fn emit(self) {
+        error!(
+            message = %format!("Receive channel closed, unable to send."),
+            error_type = error_type::COMMAND_FAILED,
+            stage = error_stage::RECEIVING,
+        );
```
Comment (neuronull) [inline src/internal_events/exec.rs:231]:
```suggestion
        let reason = "Receive channel closed, unable to send.";
        error!(
            message = reason,
            error_type = error_type::COMMAND_FAILED,
            stage = error_stage::RECEIVING,
            internal_log_rate_limit = true,
        );
```

File: src/test_util/components.rs [hunk 3]
```diff
@@ -293,6 +293,22 @@ where
         .await
 }
 
+/// Runs and returns a future and asserts that the provided test specification passes.
+#[track_caller]
+pub async fn assert_source_error<T>(
```
Comment (neuronull) [inline src/test_util/components.rs]: nice!

File: src/internal_events/exec.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (StephenWakely) [reply src/internal_events/exec.rs:241]: Yes.. doh..

[the remark this answers is not in the document]
Comment (spencergilbert) [reply src/internal_events/exec.rs:241]: @neuronull's suggestions are much better since he provided the diffs 😆

**Review by neuronull — APPROVED** [review]

**Pull request merged**